packages/server/src/fhircast/routes.ts

```diff
@@ -140,40 +118,92 @@ protectedCommonRoutes.post(
-    } else if (event['hub.event'] === 'DiagnosticReport-update') {
-      // See: https://build.fhir.org/ig/HL7/fhircast-docs/3-6-3-DiagnosticReport-update.html#:~:text=The%20Hub%20SHALL,the%20new%20updates.
-      event['context.priorVersionId'] = event['context.versionId'];
+async function handleSubscriptionRequest(req: Request, res: Response): Promise<void> {
+  const type = req.body['hub.channel.type'];
+  if (type !== 'websocket') {
+    sendOutcome(res, badRequest('Invalid hub.channel.type'));
+    return;
+  }
+
+  const mode = req.body['hub.mode'] as 'subscribe' | 'unsubscribe';
+  if (!(['subscribe', 'unsubscribe'] as const).includes(mode)) {
+    sendOutcome(res, badRequest('Invalid hub.mode'));
+    return;
+  }
+
+  const topic = req.body['hub.topic'];
+  const subscriptionEndpoint = topic; // TODO: Create separate subscription endpoint for topic
+  const config = getConfig();
+
+  switch (mode) {
+    case 'subscribe':
+      res.status(202).json({
+        'hub.channel.endpoint': getWebSocketUrl(config.baseUrl, `/ws/fhircast/${subscriptionEndpoint}`),
+      });
+      break;
+    case 'unsubscribe':
+      res.status(202).json({
+        'hub.channel.endpoint': getWebSocketUrl(config.baseUrl, `/ws/fhircast/${subscriptionEndpoint}`),
+      });
+      getRedis()
+        .publish(
+          topic,
+          JSON.stringify({
+            'hub.mode': 'denied',
+            'hub.topic': topic,
+            'hub.events': req.body['hub.events'],
+            'hub.reason': 'Subscriber unsubscribed from topic',
+          })
+        )
+        .catch((err: Error) => {
+          getLogger().error(`Error when publishing to Redis channel for FHIRcast topic: ${normalizeErrorString(err)}`, {
+            topic,
+          });
+        });
+    // break;
+  }
+}
+
+async function handleContextChangeRequest(req: Request, res: Response): Promise<void> {
+  const { event } = req.body as FhircastMessagePayload;
+  let stringifiedBody = JSON.stringify(req.body);
+  // Check if this an open event
+  if (event['hub.event'].endsWith('-open')) {
+    // TODO: Support this as a param for event type: "versionable"?
+    if (event['hub.event'] === 'DiagnosticReport-open') {
       event['context.versionId'] = generateId();
       stringifiedBody = JSON.stringify(req.body);
-      // TODO: Make sure this is actually supposed to be stored / overwrite open context? (ambiguous from docs, see: https://build.fhir.org/ig/HL7/fhircast-docs/2-9-GetCurrentContext.html)
-      await getRedis().set(`medplum:fhircast:topic:${req.params.topic}:latest`, stringifiedBody);
     }
-    await getRedis().publish(req.params.topic as string, stringifiedBody);
-    res.status(201).json({ success: true, event: body });
-  })
-);
+    // TODO: we need to get topic from event and not route param since per spec, the topic shouldn't be the slug like we have it
+    await getRedis().set(`medplum:fhircast:topic:${event['hub.topic']}:latest`, stringifiedBody);
```

Comment:
While we're making these changes, should we move forward w/ namespacing these with project ID?

This is still open to potential topic ID collision, yeah?

Reply:
Yeah for sure. We can do that in the other `FHIRcast` PR, I think I'll have to rebase that PR once this PR gets merged:
https://github.com/medplum/medplum/pull/4951